### PR TITLE
Khalil/1846 close conn rate limited peer

### PR DESCRIPTION
**Pull request by kc1116**
This PR adds a callback to the unicast rate limiters from the network middleware that will request a peer update as soon as a peer is rate limited, this will causes the underlying libp2p node to prune the connection to the rate limited peer.

### CODE REVIEW

**Review by peterargue — APPROVED** [review]

**Review by yhassanzadeh13 — COMMENTED** [review]
Looks good in general. I have some comments on the edge cases of a rate-limited node keep continuously dialing back as well as the test logic covering a rate-limited node can eventually re-connect.

File: cmd/node_builder.go [hunk 1]
```diff
@@ -251,6 +252,8 @@ type NodeConfig struct {
 	// bootstrapping options
 	SkipNwAddressBasedValidations bool
 
+	// UnicastRateLimiterDistributor notifies consumers when a peer's unicast message is rate limited.
+	UnicastRateLimiterDistributor *ratelimit.UnicastRateLimiterDistributor
```
[the remark this answers is not in the document]
Comment (kc1116) [reply cmd/node_builder.go]: https://github.com/onflow/flow-go/pull/3799/commits/8ca35d470894e79666e5118d0df050ea1b7416fb

File: cmd/scaffold.go [hunk 2]
```diff
@@ -281,21 +280,11 @@ func (fnb *FlowNodeBuilder) EnqueueNetworkInit() {
...
-			Bool(logging.KeySuspicious, true).
-			Msg("unicast peer rate limited")
-		fnb.Metrics.Network.OnRateLimitedUnicastMessage(role, msgType, topic.String(), reason.String())
-	}
+	fnb.UnicastRateLimiterDistributor = ratelimit.NewUnicastRateLimiterDistributor()
+	fnb.UnicastRateLimiterDistributor.AddConsumer(fnb.Metrics.Network)
 
 	// setup default noop unicast rate limiters
-	unicastRateLimiters := ratelimit.NewRateLimiters(ratelimit.NewNoopRateLimiter(), ratelimit.NewNoopRateLimiter(), onUnicastRateLimit, ratelimit.WithDisabledRateLimiting(fnb.BaseConfig.UnicastRateLimitDryRun))
+	unicastRateLimiters := ratelimit.NewRateLimiters(ratelimit.NewNoopRateLimiter(), ratelimit.NewNoopRateLimiter(), fnb.UnicastRateLimiterDistributor, ratelimit.WithDisabledRateLimiting(fnb.BaseConfig.UnicastRateLimitDryRun))
```
[the remark this answers is not in the document]
Comment (kc1116) [reply cmd/scaffold.go]: https://github.com/onflow/flow-go/pull/3799/commits/5969bcfb5f12d5c2dc84ce9ed9e1d060e6e02665

File: network/p2p/middleware/middleware.go [hunk 3]
```diff
@@ -840,3 +839,19 @@ func (m *Middleware) unicastMaxMsgDuration(messageType string) time.Duration {
...
+// OnRateLimitedPeer removes rate limited peer from underlying libp2pnode.
+func (m *Middleware) OnRateLimitedPeer(peerID peer.ID, role, msgType, topic, reason string) {
+	lg := m.log.With().
+		Str("peer_id", peerID.String()).
+		Str("role", role).
+		Str("message_type", msgType).
+		Str("topic", topic).
+		Str("reason", reason).
+		Bool(logging.KeySuspicious, true).Logger()
+	err := m.libP2PNode.RemovePeer(peerID)
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/middleware/middleware.go]: https://github.com/onflow/flow-go/pull/3799/commits/83cd50754aa233f88d4523047f1de7b2f538a014

File: network/test/middleware_test.go [hunk 4]
```diff
@@ -310,44 +318,50 @@ func (m *MiddlewareTestSuite) TestUnicastRateLimit_Messages() {
...
 	// wait for all rate limits before shutting down middleware
-	unittest.RequireCloseBefore(m.T(), ch, 100*time.Millisecond, "could not stop on rate limit test ch on time")
+	unittest.RequireCloseBefore(m.T(), ch, 100*time.Millisecond, "could not stop rate limit test ch on time")
+
+	// sleep for 1 seconds to allow connection pruner to prune connections
+	time.Sleep(1 * time.Second)
+	// ensure connection to rate limited peer is pruned
+	isConnected, err = libP2PNodes[0].IsConnected(expectedPID)
+	require.NoError(m.T(), err)
+	require.False(m.T(), isConnected)
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/test/middleware_test.go:342]: https://github.com/onflow/flow-go/pull/3799/commits/41aa5e12994d5a24ec3385c1fde43ee3e021dcd4

File: module/metrics/network.go [hunk 5]
```diff
@@ -334,7 +335,7 @@ func (nc *NetworkCollector) OnUnauthorizedMessage(role, msgType, topic, offense
 	nc.unAuthorizedMessagesCount.WithLabelValues(role, msgType, topic, offense).Inc()
 }
 
-// OnRateLimitedUnicastMessage tracks the number of rate limited messages seen on the network.
-func (nc *NetworkCollector) OnRateLimitedUnicastMessage(role, msgType, topic, reason string) {
+// OnRateLimitedPeer tracks the number of rate limited messages seen on the network.
+func (nc *NetworkCollector) OnRateLimitedPeer(_ peer.ID, role, msgType, topic, reason string) {
```
[the remark this answers is not in the document]
Comment (kc1116) [reply module/metrics/network.go]: The OnUnauthorizedMessage context is logged [here](https://github.com/onflow/flow-go/blob/master/network/slashing/consumer.go#L62) . I updated OnRateLimitedPeer to log context as well https://github.com/onflow/flow-go/pull/3799/commits/8bfd24893b9670fa135850ff835e5a3b07641cc5

File: network/test/middleware_test.go [hunk 6]
```diff
@@ -310,44 +326,72 @@ func (m *MiddlewareTestSuite) TestUnicastRateLimit_Messages() {
...
+
```
Comment (yhassanzadeh13) [inline network/test/middleware_test.go:343]: Please apply the followings to both `TestUnicastRateLimit_Bandwidth` and `TestUnicastRateLimit_Messages`. 

1. Please use the following approach and check both the rate-limiting peer, as well as the rate-limited peer, do not have any connections to each other. We have to check _both sides_. [code taken from here](https://github.com/onflow/flow-go/blob/f20b8c85872bd07097fefde1fadb07bd5ef3ca07/network/internal/p2pfixtures/fixtures.go#L227-L228)
    ```go
        require.Equal(t, other.Host().Network().Connectedness(thisId), network.NotConnected)
        require.Empty(t, other.Host().Network().ConnsToPeer(thisId))
    ```
    Currently, the code takes the following approach and only checks connection on the side of the rate-limited node. But we need it on both sides. 
    ```go
  	  // ensure connection to rate limited peer is pruned
	  isConnected, err = libP2PNodes[0].IsConnected(expectedPID)
	  require.NoError(m.T(), err)
	  require.False(m.T(), isConnected)
    ```
    Also, please note that for sake of testing `Connecteness` is not enough as it does not factor in the secured but not upgraded connections, so we need to also add an extra check for upgraded ones: `require.Empty(t, other.Host().Network().ConnsToPeer(thisId))`
  
  2. Please try dialing the rate-limiting peer by the rate-limited peer right after rate-limiting is done, and check that no connection is established. To do so, please use the new approach on checking the connectivity (than reading the error). You may directly use the `ensureCommunicationSilenceAmongGroups` as written [here](https://github.com/onflow/flow-go/blob/cf027cd67e9185b17249dd93a48ed749d395e47e/network/p2p/connection/connection_gater_test.go#L366) noting that both nodes should be allowed to subscribe to the `channels.PushBlocks`.

**Review by yhassanzadeh13 — APPROVED** [review]
Thanks for applying the comments. Added one comment regarding the tests and the rest looks great.

**Pull request merged**